### PR TITLE
skeleton enclave null dereference protection

### PR DESCRIPTION
**Pull request by jiazhang0**

### CODE REVIEW

**Comment by jiazhang0** [comment]
I will remove the local root author from the first commit.

**Comment by jiazhang0** [comment]
@YangLiang3 @haosanzi I have updated the latest implementation. To simplify the implementation, -z option is removed, assuming the build, signing and running platform are same. This restriction will be addressed and resolved by https://github.com/alibaba/inclavare-containers/issues/296.

File: rune/libenclave/internal/runtime/pal/skeleton/sgxsign.c [hunk 1]
```diff
@@ -295,13 +302,18 @@ static bool measure_encl(const char *path, uint8_t *mrenclave,
 	}
 
 	struct sgx_tcs *tcs = bin;
-	tcs->ssa_offset = sb.st_size;
-
+	/* SSA frame is located behind encl.bin */
+	tcs->ssa_offset = encl_offset + sb.st_size;
```
Comment (YangLiang3) [inline rune/libenclave/internal/runtime/pal/skeleton/sgxsign.c:306]: Offset of the base of the State Save Area stack, relative to the enclave base.Must be page aligned.

File: rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c [hunk 2]
```diff
@@ -687,8 +777,8 @@ int wait4child(pal_exec_args *attr)
 }
 /* *INDENT-ON* */
 
-int __pal_get_local_report(void *targetinfo, int targetinfo_len, void *report,
-			   int *report_len)
+int __pal_get_local_report(uint64_t base, void *targetinfo,
```
Comment (YangLiang3) [inline rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c:780]: Please sync parameters with api_v3_linux.go as below: other question, How and from where to pass the parameter "base"? 
static int palGetLocalReport(void *sym, void *target_info, int target_info_len,
 22                                                         void *report, int* report_len)
 23 {
 24         return ((int (*)(void *, int, void*, int*))sym)(target_info, target_info_len,
 25                                                                 report, report_len);
 26 }

**Review by YangLiang3 — COMMENTED** [review]
other LGTM

File: rune/libenclave/internal/runtime/pal/skeleton/sgxutils.c [hunk 3]
```diff
@@ -125,3 +132,72 @@ bool is_launch_control_supported(void)
...
+
+	if (mmap_min_addr) {
+		/* OOT driver cannot enable enclave dereference protection
+		 * if vm.mmap_min_addr is set to non-zero. In this case,
+		 * load_base - encl_base must equal to 0 in order to keep
+		 * being consistent between mmap area and enclave range.
+		 */
+		encl_offset = 0;
+		/* But there is no such a restriction for in-tree driver.
+		 * We have to explicitly set -z option to enable the
```
[the remark this answers is not in the document]
Comment (jiazhang0) [reply rune/libenclave/internal/runtime/pal/skeleton/sgxutils.c:173]: Change to:
```
                /* But there is no such a restriction for in-tree driver.
                 * Currently, null_dereference_protection is always true
                 * if in-tree driver is used.
                 */
```

File: rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c (hunk 2 above)
Comment (jiazhang0) [reply rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c:780]: I screwed up it. Rework.

**Comment by jiazhang0** [comment]
@YangLiang3 Here is the quick diff:

```C
diff --git a/rune/libenclave/internal/runtime/pal/skeleton/defines.h b/rune/libenclave/internal/runtime/pal/skeleton/defines.h 
index a24c01e..c137b1d 100644 
--- a/rune/libenclave/internal/runtime/pal/skeleton/defines.h 
+++ b/rune/libenclave/internal/runtime/pal/skeleton/defines.h 
@@ -34,6 +34,11 @@ bool is_oot_kernel_driver(void); 
                __tmp; \ 
        }) 
  
+#define align_up(sz, a) \ 
+       ({ \ 
+               ((sz) + (a) - 1) & ~((a) - 1)); \ 
+       }) 
+ 
 /* *INDENT-OFF* */ 
 #endif /* DEFINES_H */ 
 /* *INDENT-ON* */ 
diff --git a/rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c b/rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c
index f1815b3..cee61da 100644
--- a/rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c
+++ b/rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c
@@ -596,6 +596,7 @@ int encl_init(struct enclave_info *encl_info)
                        return -ENOENT;
        } 
  
+       bin_size = align_up(bin_size, PAGE_SIZE); 
        if (!encl_build(&secs, bin, bin_size, &sigstruct, &token, 
                        encl_info, max_mmap_size)) 
                return -EINVAL; 
@@ -777,8 +778,8 @@ int wait4child(pal_exec_args *attr) 
 } 
 /* *INDENT-ON* */ 
  
-int __pal_get_local_report(uint64_t base, void *targetinfo, 
-                          int targetinfo_len, void *report, int *report_len) 
+int __pal_get_local_report(void *targetinfo, int targetinfo_len, 
+                          void *report, int *report_len) 
 { 
        uint8_t report_data[64] = { 0, }; 
        struct sgx_report report_align;
@@ -808,7 +809,7 @@ int __pal_get_local_report(uint64_t base, void *targetinfo,
                return -1;
        }

-       ret = SGX_ENTER_3_ARGS(ECALL_REPORT, (void *) base, targetinfo,
+       ret = SGX_ENTER_3_ARGS(ECALL_REPORT, (void *) secs.base, targetinfo, 
                               report_data, &report_align); 
        if (ret) { 
                fprintf(stderr, "failed to get report\n"); 
diff --git a/rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.h b/rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.h 
index f27dfed..83c2831 100644 
--- a/rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.h 
+++ b/rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.h 
@@ -55,7 +55,7 @@ int __pal_exec(char *path, char *argv[], pal_stdio_fds *stdio, int *exit_code); 
 int __pal_create_process(pal_create_process_args *args); 
 int wait4child(pal_exec_args *attr); 
 /* *INDENT-ON* */ 
-int __pal_get_local_report(uint64_t base, void *targetinfo, int targetinfo_len, 
+int __pal_get_local_report(void *targetinfo, int targetinfo_len, 
                           void *report, int *report_len); 
 int __pal_kill(int pid, int sig); 
 int __pal_destroy(void);
diff --git a/rune/libenclave/internal/runtime/pal/skeleton/sgxsign.c b/rune/libenclave/internal/runtime/pal/skeleton/sgxsign.c
index dbe8425..22abcd2 100644
--- a/rune/libenclave/internal/runtime/pal/skeleton/sgxsign.c
+++ b/rune/libenclave/internal/runtime/pal/skeleton/sgxsign.c
        if (mmap_size % PAGE_SIZE) 
                mmap_size = (mmap_size / PAGE_SIZE + 1) * PAGE_SIZE; 
  
-       uint64_t encl_size = mmap_size; 
        uint64_t encl_offset = calc_enclave_offset(mmap_min_addr, 
                                                   !is_oot_kernel_driver()); 
-       encl_size = pow2(encl_offset + mmap_size); 
+       uint64_t encl_size = pow2(encl_offset + mmap_size); 
  
        void *bin = mmap(NULL, sb.st_size, PROT_READ | PROT_WRITE, MAP_PRIVATE, 
                         fileno(file), 0); 
@@ -302,8 +301,8 @@ static bool measure_encl(const char *path, uint8_t *mrenclave, 
        } 
  
        struct sgx_tcs *tcs = bin; 
-       /* SSA frame is located behind encl.bin */ 
-       tcs->ssa_offset = encl_offset + sb.st_size; 
+       /* SSA frame is located right behind encl.bin */
+       tcs->ssa_offset = encl_offset + align_up(sb.st_size, PAGE_SIZE);
        tcs->entry_offset += encl_offset;
        if (!mrenclave_ecreate(ctx, ssa_frame_size, encl_size))
                goto out;
diff --git a/rune/libenclave/internal/runtime/pal/skeleton/sgxutils.c b/rune/libenclave/internal/runtime/pal/skeleton/sgxutils.c 
index c01d279..1182106 100644 
--- a/rune/libenclave/internal/runtime/pal/skeleton/sgxutils.c 
+++ b/rune/libenclave/internal/runtime/pal/skeleton/sgxutils.c 
@@ -166,12 +166,12 @@ uint64_t calc_enclave_offset(uint64_t mmap_min_addr, 
                /* OOT driver cannot enable enclave dereference protection 
                 * if vm.mmap_min_addr is set to non-zero. In this case, 
                 * load_base - encl_base must equal to 0 in order to keep 
-                * being consistent between mmap area and enclave range. 
+                * consistent between mmap area and enclave range. 
                 */ 
                encl_offset = 0; 
                /* But there is no such a restriction for in-tree driver. 
-                * We have to explicitly set -z option to enable the 
-                * protection for in-tree driver. 
+                * Currently, null_dereference_protection is always true 
+                * if in-tree driver is used. 
                 */ 
                if (null_dereference_protection) 
                        encl_offset = mmap_min_addr;
```

**Review by YangLiang3 — APPROVED** [review]

**Review by haosanzi — APPROVED** [review]

**Pull request merged**